modules/dnn/src/cuda/resize.cu

```diff
@@ -0,0 +1,133 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include <cuda_runtime.h>
+#include <cuda_fp16.h>
+
+#include "math.hpp"
+#include "types.hpp"
+#include "grid_stride_range.hpp"
+#include "execution.hpp"
+
+#include "../cuda4dnn/csl/stream.hpp"
+#include "../cuda4dnn/csl/tensor.hpp"
+#include "../cuda4dnn/csl/span.hpp"
+
+#include <cuda_runtime.h>
+
+using namespace cv::dnn::cuda4dnn::csl;
+using namespace cv::dnn::cuda4dnn::csl::device;
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace kernels {
+
+    namespace raw {
+        template <class T>
+        __global__ void resize_nn(
+            Span<T> output, size_type out_height, size_type out_width,
+            View<T> input, size_type in_height, size_type in_width)
+        {
+            auto in_image_size = in_height * in_width;
+            auto out_image_size = out_height * out_width;
+
+            /* o2i = output to input */
+            auto o2i_fx = static_cast<float>(in_width) / out_width;
+            auto o2i_fy = static_cast<float>(in_height) / out_height;
+
+            /* think of the output and input as a collection of 2d images with the last axis
+             * representing the width and the last but one axis representing the height
+             *
+             * the remaining axis together form a collection of these images
+             */
+            for (auto idx : grid_stride_range(output.size())) {
+                const index_type n = idx / out_image_size;
+                const index_type x = (idx % out_image_size) % out_width;
+                const index_type y = (idx % out_image_size) / out_width;
+
+                auto in_x = static_cast<index_type>(x * o2i_fx);
+                auto in_y = static_cast<index_type>(y * o2i_fy);
```

Comment:
Perhaps we should use rounding instead of truncation.

Reply:
The OCV backend uses:

```
float input_y = y * scaleHeight;
int y0 = static_cast<int>(input_y);
```
```
float input_x = x * scaleWidth;
int x0 = static_cast<int>(input_x);
```

What kind of rounding should be used? I think it should be round to nearest with `x.5` rounded down.

Reply:
After digging through the code, I believe that `cv::resize` uses floor for rounding in NN method.

Since the numbers are always positive, truncation and floor would give the same effect. Truncation is also faster than rounding.

Reply:
Right!